nifi-nar-bundles/nifi-azure-bundle/nifi-azure-processors/src/main/java/org/apache/nifi/processors/azure/cosmos/document/GetAzureCosmosDBRecord.java

```diff
@@ -0,0 +1,266 @@
+        this.writerFactory =context.getProperty(WRITER_FACTORY).asControllerService(RecordSetWriterFactory.class);
+
+        final FlowFile input = context.hasIncomingConnection() ? session.get() : null;
+        logger = getLogger();
+
+        if (input == null && context.hasNonLoopConnection()) {
+            if (logger.isDebugEnabled()) {
+                logger.debug("No flowfile input and NonLoopConnection. Ending onTrigger... ");
+            }
+            return;
+        }
+        final String query = getQuery(context, session, input);
+        final Map<String, String> attributes = getAttributes(context, input);
+        if (logger.isDebugEnabled()) {
+            logger.debug("Running Cosmos SQL query : " + query);
+        }
+        final CosmosQueryRequestOptions queryOptions = new CosmosQueryRequestOptions();
+        final CosmosContainer container = getContainer();
+        final CosmosPagedIterable<JsonNode> response =  container != null
+                ? container.queryItems(query, queryOptions, JsonNode.class)
+                : null;
+        if (response == null) {
+            logger.error("Fails to get CosmosPagedIterable<JsonNode> response");
+            if (input != null) {
+                session.transfer(input, REL_FAILURE);
+            }
+            return;
+        }
+        FlowFile output = input != null ? session.create(input) : session.create();
+        try {
+            logger.debug("Start to process data from Azure Cosmos DB");
+            final String schemaName = context.getProperty(SCHEMA_NAME).evaluateAttributeExpressions(input).getValue();
+            try (OutputStream out = session.write(output)) {
+                Map<String, String> attrs = input != null ? input.getAttributes() : new HashMap<String, String>(){{
+                    put("schema.name", schemaName);
+                }};
+                RecordSchema schema = writerFactory.getSchema(attrs, null);
+                RecordSetWriter writer = writerFactory.createWriter(getLogger(), schema, out, attrs);
+                final AtomicLong count = new AtomicLong();
+                writer.beginRecordSet();
+
+                response.forEach(data ->{
+                    try {
+                        Map<String,Object> mapObj = mapper.convertValue(data, new TypeReference<Map<String, Object>>(){});
+                        Record record = new MapRecord(schema, mapObj);
+                        writer.write(record);
+                    } catch(IOException | IllegalArgumentException ex) {
+                        throw new RuntimeException(ex);
+                    }
+                    count.incrementAndGet();
+                });
+                writer.finishRecordSet();
+                writer.close();
+                out.close();
+                attributes.put("record.count", String.valueOf(count.get()));
+            } catch (SchemaNotFoundException e) {
+                throw new RuntimeException(e);
+            }
```

Comment:
I'd recommend a little refactor here, to break down the complexity of this method. I'd extract the nested try block and the map initialization.

```suggestion
    private void processSession(ProcessSession session, FlowFile input, Map<String, String> attributes, CosmosPagedIterable<JsonNode> response, FlowFile output, String schemaName) throws IOException {
        final Map<String, String> attributeMap = initAttributeMap(input, schemaName);
        try (OutputStream out = session.write(output)) {
            RecordSchema schema = writerFactory.getSchema(attributeMap, null);
            RecordSetWriter writer = writerFactory.createWriter(getLogger(), schema, out, attributeMap);
            final AtomicLong count = new AtomicLong();
            writer.beginRecordSet();

            response.forEach(data -> processData(schema, writer, count, data));
            writer.finishRecordSet();
            writer.close();
            attributes.put("record.count", String.valueOf(count.get()));
        } catch (SchemaNotFoundException e) {
            throw new RuntimeException(e);
        }
    }
    
    private void processData(RecordSchema schema, RecordSetWriter writer, AtomicLong count, JsonNode data) {
        try {
            Map<String,Object> mapObj = mapper.convertValue(data, new TypeReference<Map<String, Object>>(){});
            Record record = new MapRecord(schema, mapObj);
            writer.write(record);
        } catch(IOException | IllegalArgumentException ex) {
            throw new RuntimeException(ex);
        }
        count.incrementAndGet();
    }
```